examples/common/convert_c4.py

```diff
@@ -4,105 +4,359 @@
+                    buffer['text'] = buffer.get('text', '') + self.sep_text + sample
+                    while len(buffer['text']) >= self.max_length:
+                        concat_sample = {}
+                        concat_sample['text'] = buffer['text'][:self.
+                                                                max_length]
+                        buffer['text'] = buffer['text'][self.max_length:]
+
+                        yield {
+                            'text': concat_sample['text'].encode('utf-8')
+                        }
+                        n_samples += 1
+                        if n_samples == self.expected_num_samples:
+                            return
+
+
+class TokenizedC4(ShardedC4):
+    """A tokenized, concatenated, iterable dataset.
+
+    To use data created by this class and written to MDS format:
+
+    ```python
+        import torch
+        from streaming.base import StreamingDataset
+        from transformers import AutoTokenizer
+
+        tokenizer = AutoTokenizer.from_pretrained('your/tokenizer')
+        ds = StreamingDataset(local='mds-data-folder', split='val')
+
+        # note, you need to copy the numpy array because the original is non-writeable
+        # and torch does not support non-writeable tensors, so you get a scary warning and
+        # if you do try to write to the tensor you get undefined behavior
+        tokens = torch.from_numpy(np.frombuffer(ds[0]['tokens'], dtype=np.int64).copy())
+        print(tokenizer.decode(tokens))
+    ```
+    """
+
+    def __init__(self, tokenizer: PreTrainedTokenizerBase, *args, **kwargs):
+        self.tokenizer = tokenizer
+        super().__init__(*args, **kwargs)
+
+    def generate_samples(self) -> Iterable[Dict[str, bytes]]:
+        """Generator over each dataset sample.
+
+        Args:
+            samples (IterableDataset): An iterable dataset that is multi-worker compatible
+
+        Yields:
+            Sample dicts.
+        """
+
+        n_samples = 0
+        buffer = {}
+        while True:
+            for batch in self.loader:
+                encoded = self.tokenizer(batch['text'], truncation=False, padding=False)
+                for iids in encoded['input_ids']:
+                    buffer['tokens'] = buffer.get('tokens', []) + iids
+                    while len(buffer['tokens']) >= self.max_length:
+                        concat_sample = {}
+                        concat_sample['tokens'] = buffer['tokens'][:self.
+                                                                max_length]
+                        buffer['tokens'] = buffer['tokens'][self.max_length:]
+
+                        yield {
+                            # convert to bytes to store in MDS binary format
+                            'tokens': np.asarray(concat_sample['tokens']).tobytes()
+                        }
+                        n_samples += 1
+                        if n_samples == self.expected_num_samples:
```

Comment:
do we need to dump whatever is left in the buffer at the end?

Reply:
I was assuming it would get GC'ed but I can dump it explicitly

Reply:
er, by "dump" i meant "yield"

Reply:
I don't think so. We've reached the "end"

Reply:
sorry this was the most confusing comment. what I _actually_ meant was, is there a case where we don't return at this statement, but instead complete the for loop, and then we should yield the last partial sample?

Reply:
I don't think we ever want to yield the partial sample, because the user specified how long each sample should be, and it won't be that long (well, 1 in max_length times I guess it will be)

Reply:
got it, sounds good. sorry for the back and forth on this one :)

Reply:
no worries